// @dcc-commons-ng/services/esm5/lib/route/route.service.js
/*
 * Copyright (c) 2019 ELM. All rights reserved.
 */
import { animate, animateChild, group, query, sequence, style, transition, trigger } from '@angular/animations';
export var RouteAnimations = trigger('routerAnimations', [
    transition('* <=> *', [
        // css styles at start of transition
        //style({ opacity: 0 }),
        // animation and styles at end of transition
        //animate('0.5s', style({ opacity: 1 }))
        query(':enter, :leave', style({ position: 'fixed', width: '100%' }), { optional: true }),
        query(':enter', style({ transform: 'translateX(100%)' }), { optional: true }),
        sequence([
            query(':leave', animateChild(), { optional: true }),
            group([
                query(':leave', [
                    style({ transform: 'translateX(0%)' }),
                    animate('500ms cubic-bezier(.75,-0.48,.26,1.52)', style({ transform: 'translateX(-100%)' }))
                ], { optional: true }),
                query(':enter', [
                    style({ transform: 'translateX(100%)' }),
                    animate('500ms cubic-bezier(.75,-0.48,.26,1.52)', style({ transform: 'translateX(0%)' }))
                ], { optional: true }),
            ]),
            query(':enter', animateChild(), { optional: true }),
        ])
    ])
]);
/**
 * Provides helper methods to create routes.
 */
var Route = /** @class */ (function () {
    function Route() {
    }
    /**
     * Creates routes using the shell component and authentication.
     * @param routes The routes to add.
     * @return {Route} The new route using shell as the base.
     */
    Route.withShell = function (routes, appLayout, authGuard) {
        return {
            path: '',
            component: appLayout,
            children: routes,
            canActivate: [authGuard],
            // Reuse ShellComponent instance when navigating between child views
            data: { reuse: true }
        };
    };
    return Route;
}());
export { Route };
