// @dcc-commons-ng/services/esm2015/lib/services.module.js
/*
 * Copyright (c) 2019 ELM. All rights reserved.
 */
var DccCommonsNgServicesModule_1;
import { __decorate, __metadata, __param } from "tslib";
import { NgModule, Optional, SkipSelf } from '@angular/core';
import { CommonModule } from '@angular/common';
import { HttpClient, HttpClientModule } from '@angular/common/http';
import { RouteReuseStrategy, RouterModule } from '@angular/router';
import { TranslateLoader, TranslateModule } from '@ngx-translate/core';
import { RouteReusableStrategy } from './route/route-reusable-strategy';
import { I18nService } from './i18n/i18n.service';
import { HttpService } from './http/http.service';
import { HttpCacheService } from './http/http-cache.service';
import { ApiPrefixInterceptor } from './http/api-prefix.interceptor';
import { ErrorHandlerInterceptor } from './http/error-handler.interceptor';
import { CacheInterceptor } from './http/cache.interceptor';
import { JwtInterceptor, LOGIN_URL, PUBLIC_URL_PATTERNS } from './http/jwt.interceptor';
import { TranslateHttpLoader } from "@ngx-translate/http-loader";
import { CookieService } from "ngx-cookie-service";
import { CsrfInterceptor } from "./http/csrf.interceptor";
import { MockInterceptor } from "./http/mock.interceptor";
export const I18N_FOLDER_PATH = './assets/i18n/';
const ɵ0 = (HttpLoaderFactory);
let DccCommonsNgServicesModule = DccCommonsNgServicesModule_1 = class DccCommonsNgServicesModule {
    constructor(parentModule) {
        // Import guard
        if (parentModule) {
            throw new Error(`${parentModule} has already been loaded. Import Services module in the AppModule only.`);
        }
    }
    static forRoot(publicUrlPatterns = [], loginUrl, i18nFolderPath) {
        return {
            ngModule: DccCommonsNgServicesModule_1,
            providers: [
                { provide: PUBLIC_URL_PATTERNS, useValue: publicUrlPatterns },
                { provide: LOGIN_URL, useValue: loginUrl },
                { provide: I18N_FOLDER_PATH, useValue: i18nFolderPath }
            ]
        };
    }
};
DccCommonsNgServicesModule.ctorParameters = () => [
    { type: DccCommonsNgServicesModule, decorators: [{ type: Optional }, { type: SkipSelf }] }
];
__decorate([
    __param(0, Optional()), __param(1, Optional()), __param(2, Optional()),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Array, String, String]),
    __metadata("design:returntype", Object)
], DccCommonsNgServicesModule, "forRoot", null);
DccCommonsNgServicesModule = DccCommonsNgServicesModule_1 = __decorate([
    NgModule({
        imports: [
            CommonModule,
            HttpClientModule,
            TranslateModule.forRoot({
                loader: {
                    provide: TranslateLoader,
                    useFactory: ɵ0,
                    deps: [HttpClient, I18N_FOLDER_PATH]
                }
            }),
            RouterModule
        ],
        declarations: [],
        providers: [
            I18nService,
            HttpCacheService,
            ApiPrefixInterceptor,
            ErrorHandlerInterceptor,
            CacheInterceptor,
            CookieService,
            JwtInterceptor,
            CsrfInterceptor,
            MockInterceptor,
            {
                provide: HttpClient,
                useClass: HttpService
            },
            {
                provide: RouteReuseStrategy,
                useClass: RouteReusableStrategy
            }
        ]
    }),
    __param(0, Optional()), __param(0, SkipSelf()),
    __metadata("design:paramtypes", [DccCommonsNgServicesModule])
], DccCommonsNgServicesModule);
export { DccCommonsNgServicesModule };
export function HttpLoaderFactory(http, i18nFolderPath) {
    return new TranslateHttpLoader(http, i18nFolderPath, '.json');
}
export { ɵ0 };
